Ah, now there are different failures:

```
2022-01-05T09:51:18.0424204Z 2022-01-05 09:51:17,886 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (vert.x-worker-thread-0) Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
2022-01-05T09:51:18.0427462Z 	[error]: Build step io.quarkus.resteasy.deployment.ResteasyBuiltinsProcessor#addStaticResourcesExceptionMapper threw an exception: java.nio.file.ClosedFileSystemException
2022-01-05T09:51:18.0430453Z 	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.ensureOpen(ZipFileSystem.java:1146)
2022-01-05T09:51:18.0431947Z 	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.checkAccess(ZipFileSystem.java:362)
2022-01-05T09:51:18.0433439Z 	at jdk.zipfs/jdk.nio.zipfs.ZipPath.checkAccess(ZipPath.java:843)
2022-01-05T09:51:18.0434975Z 	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.checkAccess(ZipFileSystemProvider.java:185)
2022-01-05T09:51:18.0436962Z 	at java.base/java.nio.file.Files.exists(Files.java:2440)
2022-01-05T09:51:18.0438200Z 	at io.quarkus.paths.PathList.resolveExistingOrNull(PathList.java:121)
2022-01-05T09:51:18.0440039Z 	at io.quarkus.deployment.ApplicationArchive.getChildPath(ApplicationArchive.java:93)
2022-01-05T09:51:18.0443454Z 	at io.quarkus.resteasy.deployment.ResteasyBuiltinsProcessor.lambda$addStaticResourcesExceptionMapper$0(ResteasyBuiltinsProcessor.java:120)
2022-01-05T09:51:18.0446285Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
2022-01-05T09:51:18.0447679Z 	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1621)
2022-01-05T09:51:18.0449132Z 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
2022-01-05T09:51:18.0450924Z 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
2022-01-05T09:51:18.0452620Z 	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
2022-01-05T09:51:18.0454278Z 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
2022-01-05T09:51:18.0455876Z 	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
2022-01-05T09:51:18.0460089Z 	at io.quarkus.resteasy.deployment.ResteasyBuiltinsProcessor.addStaticResourcesExceptionMapper(ResteasyBuiltinsProcessor.java:123)
2022-01-05T09:51:18.0463432Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2022-01-05T09:51:18.0465566Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2022-01-05T09:51:18.0468015Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2022-01-05T09:51:18.0469850Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2022-01-05T09:51:18.0471233Z 	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:887)
2022-01-05T09:51:18.0472477Z 	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
2022-01-05T09:51:18.0473634Z 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
2022-01-05T09:51:18.0474997Z 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
2022-01-05T09:51:18.0476628Z 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
2022-01-05T09:51:18.0477829Z 	at java.base/java.lang.Thread.run(Thread.java:829)
2022-01-05T09:51:18.0478906Z 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
2022-01-05T09:51:18.0479986Z 
2022-01-05T09:51:18.0539422Z 	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:330)
2022-01-05T09:51:18.0542324Z 	at io.quarkus.runner.bootstrap.AugmentActionImpl.reloadExistingApplication(AugmentActionImpl.java:265)
2022-01-05T09:51:18.0545928Z 	at io.quarkus.runner.bootstrap.AugmentActionImpl.reloadExistingApplication(AugmentActionImpl.java:60)
2022-01-05T09:51:18.0548842Z 	at io.quarkus.deployment.dev.IsolatedDevModeMain.restartApp(IsolatedDevModeMain.java:231)
2022-01-05T09:51:18.0551505Z 	at io.quarkus.deployment.dev.IsolatedDevModeMain.restartCallback(IsolatedDevModeMain.java:214)
2022-01-05T09:51:18.0554064Z 	at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.doScan(RuntimeUpdatesProcessor.java:522)
2022-01-05T09:51:18.0556795Z 	at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.doScan(RuntimeUpdatesProcessor.java:423)
2022-01-05T09:51:18.0560734Z 	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$4.handle(VertxHttpHotReplacementSetup.java:152)
2022-01-05T09:51:18.0564270Z 	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$4.handle(VertxHttpHotReplacementSetup.java:139)
2022-01-05T09:51:18.0566782Z 	at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:159)
2022-01-05T09:51:18.0568127Z 	at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
2022-01-05T09:51:18.0569534Z 	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:157)
2022-01-05T09:51:18.0571107Z 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
2022-01-05T09:51:18.0572458Z 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
2022-01-05T09:51:18.0574204Z 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
2022-01-05T09:51:18.0575882Z 	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
2022-01-05T09:51:18.0578429Z 	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
2022-01-05T09:51:18.0580749Z 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
2022-01-05T09:51:18.0582301Z 	at java.base/java.lang.Thread.run(Thread.java:829)
2022-01-05T09:51:18.0583531Z Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
2022-01-05T09:51:18.0586557Z 	[error]: Build step io.quarkus.resteasy.deployment.ResteasyBuiltinsProcessor#addStaticResourcesExceptionMapper threw an exception: java.nio.file.ClosedFileSystemException
2022-01-05T09:51:18.0589908Z 	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.ensureOpen(ZipFileSystem.java:1146)
2022-01-05T09:51:18.0591719Z 	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.checkAccess(ZipFileSystem.java:362)
2022-01-05T09:51:18.0594002Z 	at jdk.zipfs/jdk.nio.zipfs.ZipPath.checkAccess(ZipPath.java:843)
2022-01-05T09:51:18.0596954Z 	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.checkAccess(ZipFileSystemProvider.java:185)
2022-01-05T09:51:18.0598586Z 	at java.base/java.nio.file.Files.exists(Files.java:2440)
2022-01-05T09:51:18.0599916Z 	at io.quarkus.paths.PathList.resolveExistingOrNull(PathList.java:121)
2022-01-05T09:51:18.0762485Z 	at io.quarkus.deployment.ApplicationArchive.getChildPath(ApplicationArchive.java:93)
2022-01-05T09:51:18.0765529Z 	at io.quarkus.resteasy.deployment.ResteasyBuiltinsProcessor.lambda$addStaticResourcesExceptionMapper$0(ResteasyBuiltinsProcessor.java:120)
2022-01-05T09:51:18.0768343Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
2022-01-05T09:51:18.0769970Z 	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1621)
2022-01-05T09:51:18.0771551Z 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
2022-01-05T09:51:18.0773459Z 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
2022-01-05T09:51:18.0775359Z 	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
2022-01-05T09:51:18.0776899Z 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
2022-01-05T09:51:18.0778538Z 	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
2022-01-05T09:51:18.0789036Z 	at io.quarkus.resteasy.deployment.ResteasyBuiltinsProcessor.addStaticResourcesExceptionMapper(ResteasyBuiltinsProcessor.java:123)
2022-01-05T09:51:18.0792564Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2022-01-05T09:51:18.0795147Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2022-01-05T09:51:18.0802295Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2022-01-05T09:51:18.0804411Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2022-01-05T09:51:18.0805817Z 	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:887)
2022-01-05T09:51:18.0807921Z 	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
2022-01-05T09:51:18.0809212Z 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
2022-01-05T09:51:18.0811035Z 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
2022-01-05T09:51:18.0812703Z 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
2022-01-05T09:51:18.0813943Z 	at java.base/java.lang.Thread.run(Thread.java:829)
2022-01-05T09:51:18.0814870Z 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
2022-01-05T09:51:18.0815578Z 
2022-01-05T09:51:18.0816261Z 	at io.quarkus.builder.Execution.run(Execution.java:116)
2022-01-05T09:51:18.0817736Z 	at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:79)
2022-01-05T09:51:18.0819453Z 	at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:160)
2022-01-05T09:51:18.0821299Z 	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:328)
2022-01-05T09:51:18.0822539Z 	... 18 more
2022-01-05T09:51:18.0823285Z Caused by: java.nio.file.ClosedFileSystemException
2022-01-05T09:51:18.0824555Z 	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.ensureOpen(ZipFileSystem.java:1146)
2022-01-05T09:51:18.0825993Z 	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.checkAccess(ZipFileSystem.java:362)
2022-01-05T09:51:18.0827282Z 	at jdk.zipfs/jdk.nio.zipfs.ZipPath.checkAccess(ZipPath.java:843)
2022-01-05T09:51:18.0828827Z 	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.checkAccess(ZipFileSystemProvider.java:185)
2022-01-05T09:51:18.0830265Z 	at java.base/java.nio.file.Files.exists(Files.java:2440)
2022-01-05T09:51:18.0831474Z 	at io.quarkus.paths.PathList.resolveExistingOrNull(PathList.java:121)
2022-01-05T09:51:18.0833314Z 	at io.quarkus.deployment.ApplicationArchive.getChildPath(ApplicationArchive.java:93)
2022-01-05T09:51:18.0836191Z 	at io.quarkus.resteasy.deployment.ResteasyBuiltinsProcessor.lambda$addStaticResourcesExceptionMapper$0(ResteasyBuiltinsProcessor.java:120)
2022-01-05T09:51:18.0838771Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
2022-01-05T09:51:18.0840323Z 	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1621)
2022-01-05T09:51:18.0842360Z 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
2022-01-05T09:51:18.0844079Z 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
2022-01-05T09:51:18.0845865Z 	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
2022-01-05T09:51:18.0847387Z 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
2022-01-05T09:51:18.0849033Z 	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
2022-01-05T09:51:18.0852207Z 	at io.quarkus.resteasy.deployment.ResteasyBuiltinsProcessor.addStaticResourcesExceptionMapper(ResteasyBuiltinsProcessor.java:123)
2022-01-05T09:51:18.0855803Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2022-01-05T09:51:18.0857925Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2022-01-05T09:51:18.0861809Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2022-01-05T09:51:18.0863766Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2022-01-05T09:51:18.0865255Z 	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:887)
2022-01-05T09:51:18.0866715Z 	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
2022-01-05T09:51:18.0867908Z 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
2022-01-05T09:51:18.0869330Z 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
2022-01-05T09:51:18.0871002Z 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
2022-01-05T09:51:18.0872243Z 	at java.base/java.lang.Thread.run(Thread.java:829)
2022-01-05T09:51:18.0873445Z 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
```

--- a/extensions/micrometer/deployment/src/test/java/io/quarkus/micrometer/runtime/MicrometerCounterInterceptorTest.java
+++ b/extensions/micrometer/deployment/src/test/java/io/quarkus/micrometer/runtime/MicrometerCounterInterceptorTest.java
@@ -18,6 +18,7 @@ import io.quarkus.micrometer.test.CountedResource;
 import io.quarkus.micrometer.test.GuardedResult;
 import io.quarkus.micrometer.test.TimedResource;
 import io.quarkus.test.QuarkusUnitTest;
+import io.smallrye.mutiny.Uni;
 
 public class MicrometerCounterInterceptorTest {
 
@@ -168,6 +169,85 @@ public class MicrometerCounterInterceptorTest {
         CompletableFuture<?> completableFuture = counted.emptyAsyncMetricName(guardedResult);
         guardedResult.complete(new NullPointerException());
         Assertions.assertThrows(java.util.concurrent.CompletionException.class, () -> completableFuture.join());
+
+        Counter counter = registry.get("method.counted")
+                .tag("method", "emptyMetricName")
+                .tag("class", "io.quarkus.micrometer.test.CountedResource")
+                .tag("exception", "NullPointerException")
+                .tag("result", "failure").counter();
+        Assertions.assertNotNull(counter);
+        Assertions.assertEquals(1, counter.count());
+    }
+
+    @Test
+    void testCountUniFailuresOnly_NoMetricsOnSuccess() {
+        GuardedResult guardedResult = new GuardedResult();
+        Uni<?> uni = counted.onlyCountUniFailures(guardedResult);
+        guardedResult.complete();
+        uni.subscribe().asCompletionStage().join();
+
+        Assertions.assertThrows(MeterNotFoundException.class, () -> registry.get("uni.none").counter());
+    }
+
+    @Test
+    void testCountUniAllMetrics_MetricsOnSuccess() {
+        GuardedResult guardedResult = new GuardedResult();
+        Uni<?> uni = counted.countAllUniInvocations(guardedResult);
+        guardedResult.complete();
+        uni.subscribe().asCompletionStage().join();
+
+        Counter counter = registry.get("uni.all")
+                .tag("method", "countAllUniInvocations")
+                .tag("class", "io.quarkus.micrometer.test.CountedResource")
+                .tag("extra", "tag")
+                .tag("exception", "none")
+                .tag("result", "success").counter();
+        Assertions.assertNotNull(counter);
+        Assertions.assertEquals(1, counter.count());
+    }
+
+    @Test
+    void testCountUniAllMetrics_MetricsOnFailure() {
+        GuardedResult guardedResult = new GuardedResult();
+        Uni<?> uni = counted.countAllUniInvocations(guardedResult);
+        guardedResult.complete(new NullPointerException());
+        Assertions.assertThrows(java.util.concurrent.CompletionException.class,
+                () -> uni.subscribe().asCompletionStage().join());
+
+        Counter counter = registry.get("uni.all")
+                .tag("method", "countAllUniInvocations")
+                .tag("class", "io.quarkus.micrometer.test.CountedResource")
+                .tag("extra", "tag")
+                .tag("exception", "NullPointerException")
+                .tag("result", "failure").counter();
+        Assertions.assertNotNull(counter);
+        Assertions.assertEquals(1, counter.count());
+        Assertions.assertNull(counter.getId().getDescription());
+    }
+
+    @Test
+    void testCountUniEmptyMetricName_Success() {
+        GuardedResult guardedResult = new GuardedResult();
+        Uni<?> uni = counted.emptyUniMetricName(guardedResult);
+        guardedResult.complete();
+        uni.subscribe().asCompletionStage().join();
+
+        Counter counter = registry.get("method.counted")
+                .tag("method", "emptyUniMetricName")
+                .tag("class", "io.quarkus.micrometer.test.CountedResource")
+                .tag("exception", "none")
+                .tag("result", "success").counter();
+        Assertions.assertNotNull(counter);
+        Assertions.assertEquals(1, counter.count());
+    }
+
+    @Test
+    void testCountUniEmptyMetricName_Failure() {
+        GuardedResult guardedResult = new GuardedResult();
+        Uni<?> uni = counted.emptyUniMetricName(guardedResult);
+        guardedResult.complete(new NullPointerException());
+        Assertions.assertThrows(java.util.concurrent.CompletionException.class,
+                () -> uni.subscribe().asCompletionStage().join());
 
         Counter counter = registry.get("method.counted")
                 .tag("method", "emptyMetricName")

--- a/extensions/micrometer/deployment/src/test/java/io/quarkus/micrometer/runtime/MicrometerTimedInterceptorTest.java
+++ b/extensions/micrometer/deployment/src/test/java/io/quarkus/micrometer/runtime/MicrometerTimedInterceptorTest.java
@@ -18,6 +18,7 @@ import io.quarkus.micrometer.test.CountedResource;
 import io.quarkus.micrometer.test.GuardedResult;
 import io.quarkus.micrometer.test.TimedResource;
 import io.quarkus.test.QuarkusUnitTest;
+import io.smallrye.mutiny.Uni;
 
 public class MicrometerTimedInterceptorTest {
     @RegisterExtension
@@ -100,6 +101,39 @@ public class MicrometerTimedInterceptorTest {
     }
 
     @Test
+    void testTimeMethod_Uni() {
+        GuardedResult guardedResult = new GuardedResult();
+        Uni<?> uni = timed.uniCall(guardedResult);
+        guardedResult.complete();
+        uni.subscribe().asCompletionStage().join();
+
+        Timer timer = registry.get("uni.call")
+                .tag("method", "uniCall")
+                .tag("class", "io.quarkus.micrometer.test.TimedResource")
+                .tag("exception", "none")
+                .tag("extra", "tag").timer();
+        Assertions.assertNotNull(timer);
+        Assertions.assertEquals(1, timer.count());
+    }
+
+    @Test
+    void testTimeMethod_UniFailed() {
+        GuardedResult guardedResult = new GuardedResult();
+        Uni<?> uni = timed.uniCall(guardedResult);
+        guardedResult.complete(new NullPointerException());
+        Assertions.assertThrows(java.util.concurrent.CompletionException.class,
+                () -> uni.subscribe().asCompletionStage().join());
+
+        Timer timer = registry.get("uni.call")
+                .tag("method", "uniCall")
+                .tag("class", "io.quarkus.micrometer.test.TimedResource")
+                .tag("exception", "NullPointerException")
+                .tag("extra", "tag").timer();
+        Assertions.assertNotNull(timer);
+        Assertions.assertEquals(1, timer.count());
+    }
+
+    @Test
     void testTimeMethod_LongTaskTimer() {
         timed.longCall(false);
         LongTaskTimer timer = registry.get("longCall")
@@ -146,6 +180,37 @@ public class MicrometerTimedInterceptorTest {
 
         LongTaskTimer timer = registry.get("async.longCall")
                 .tag("method", "longAsyncCall")
+                .tag("class", "io.quarkus.micrometer.test.TimedResource")
+                .tag("extra", "tag").longTaskTimer();
+        Assertions.assertNotNull(timer);
+        Assertions.assertEquals(0, timer.activeTasks());
+    }
+
+    @Test
+    void testTimeMethod_LongTaskTimer_Uni() {
+        GuardedResult guardedResult = new GuardedResult();
+        Uni<?> uni = timed.longUniCall(guardedResult);
+        guardedResult.complete();
+        uni.subscribe().asCompletionStage().join();
+
+        LongTaskTimer timer = registry.get("uni.longCall")
+                .tag("method", "longUniCall")
+                .tag("class", "io.quarkus.micrometer.test.TimedResource")
+                .tag("extra", "tag").longTaskTimer();
+        Assertions.assertNotNull(timer);
+        Assertions.assertEquals(0, timer.activeTasks());
+    }
+
+    @Test
+    void testTimeMethod_LongTaskTimer_UniFailed() {
+        GuardedResult guardedResult = new GuardedResult();
+        Uni<?> uni = timed.longUniCall(guardedResult);
+        guardedResult.complete(new NullPointerException());
+        Assertions.assertThrows(java.util.concurrent.CompletionException.class,
+                () -> uni.subscribe().asCompletionStage().join());
+
+        LongTaskTimer timer = registry.get("uni.longCall")
+                .tag("method", "longUniCall")
                 .tag("class", "io.quarkus.micrometer.test.TimedResource")
                 .tag("extra", "tag").longTaskTimer();
         Assertions.assertNotNull(timer);

--- a/extensions/micrometer/deployment/src/test/java/io/quarkus/micrometer/test/CountedResource.java
+++ b/extensions/micrometer/deployment/src/test/java/io/quarkus/micrometer/test/CountedResource.java
@@ -7,6 +7,7 @@ import java.util.concurrent.CompletableFuture;
 import javax.enterprise.context.ApplicationScoped;
 
 import io.micrometer.core.annotation.Counted;
+import io.smallrye.mutiny.Uni;
 
 @ApplicationScoped
 public class CountedResource {
@@ -41,6 +42,21 @@ public class CountedResource {
     @Counted
     public CompletableFuture<?> emptyAsyncMetricName(GuardedResult guardedResult) {
         return supplyAsync(guardedResult::get);
+    }
+
+    @Counted(value = "uni.none", recordFailuresOnly = true)
+    public Uni<?> onlyCountUniFailures(GuardedResult guardedResult) {
+        return Uni.createFrom().item(guardedResult::get);
+    }
+
+    @Counted(value = "uni.all", extraTags = { "extra", "tag" })
+    public Uni<?> countAllUniInvocations(GuardedResult guardedResult) {
+        return Uni.createFrom().item(guardedResult::get);
+    }
+
+    @Counted
+    public Uni<?> emptyUniMetricName(GuardedResult guardedResult) {
+        return Uni.createFrom().item(guardedResult::get);
     }
 
 }

--- a/extensions/micrometer/deployment/src/test/java/io/quarkus/micrometer/test/TimedResource.java
+++ b/extensions/micrometer/deployment/src/test/java/io/quarkus/micrometer/test/TimedResource.java
@@ -2,11 +2,14 @@ package io.quarkus.micrometer.test;
 
 import static java.util.concurrent.CompletableFuture.supplyAsync;
 
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
 import java.util.concurrent.CompletableFuture;
 
 import javax.enterprise.context.ApplicationScoped;
 
 import io.micrometer.core.annotation.Timed;
+import io.smallrye.mutiny.Uni;
 
 @ApplicationScoped
 public class TimedResource {
@@ -34,6 +37,11 @@ public class TimedResource {
         return supplyAsync(guardedResult::get);
     }
 
+    @Timed(value = "uni.call", extraTags = { "extra", "tag" })
+    public Uni<?> uniCall(GuardedResult guardedResult) {
+        return Uni.createFrom().item(guardedResult::get);
+    }
+
     @Timed(value = "async.longCall", extraTags = { "extra", "tag" }, longTask = true)
     public CompletableFuture<?> longAsyncCall(GuardedResult guardedResult) {
         try {
@@ -41,6 +49,11 @@ public class TimedResource {
         } catch (InterruptedException e) {
         }
         return supplyAsync(guardedResult::get);
+    }
+
+    @Timed(value = "uni.longCall", extraTags = { "extra", "tag" }, longTask = true)
+    public Uni<?> longUniCall(GuardedResult guardedResult) {
+        return Uni.createFrom().item(guardedResult::get).onItem().delayIt().by(Duration.of(3, ChronoUnit.MILLIS));
     }
 
     @Timed(value = "alpha", extraTags = { "extra", "tag" })

--- a/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/MicrometerCountedInterceptor.java
+++ b/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/MicrometerCountedInterceptor.java
@@ -2,6 +2,7 @@ package io.quarkus.micrometer.runtime;
 
 import java.lang.reflect.Method;
 import java.util.concurrent.CompletionStage;
+import java.util.function.BiConsumer;
 
 import javax.annotation.Priority;
 import javax.interceptor.AroundInvoke;
@@ -12,6 +13,8 @@ import io.micrometer.core.instrument.Counter;
 import io.micrometer.core.instrument.MeterRegistry;
 import io.micrometer.core.instrument.Tags;
 import io.quarkus.arc.ArcInvocationContext;
+import io.smallrye.mutiny.Uni;
+import io.smallrye.mutiny.tuples.Functions;
 
 /**
  * Quarkus declared interceptor responsible for intercepting all methods
@@ -51,6 +54,7 @@ public class MicrometerCountedInterceptor {
      * @throws Throwable When the intercepted method throws one.
      */
     @AroundInvoke
+    @SuppressWarnings("unchecked")
     Object countedMethod(ArcInvocationContext context) throws Exception {
         MicrometerCounted counted = context.findIterceptorBinding(MicrometerCounted.class);
         if (counted == null) {
@@ -59,13 +63,27 @@ public class MicrometerCountedInterceptor {
         Method method = context.getMethod();
         Tags commonTags = getCommonTags(method.getDeclaringClass().getName(), method.getName());
 
-        // If we're working with a CompletionStage
-        final boolean stopWhenCompleted = CompletionStage.class.isAssignableFrom(method.getReturnType());
-        if (stopWhenCompleted) {
+        Class<?> returnType = method.getReturnType();
+        if (TypesUtil.isCompletionStage(returnType)) {
             try {
-                return ((CompletionStage<?>) context.proceed()).whenComplete((result, throwable) -> {
-                    recordCompletionResult(counted, commonTags, throwable);
+                return ((CompletionStage<?>) context.proceed()).whenComplete(new BiConsumer<Object, Throwable>() {
+                    @Override
+                    public void accept(Object o, Throwable throwable) {
+                        recordCompletionResult(counted, commonTags, throwable);
+                    }
                 });
+            } catch (Throwable e) {
+                record(counted, commonTags, e);
+            }
+        } else if (TypesUtil.isUni(returnType)) {
+            try {
+                return ((Uni<Object>) context.proceed()).onTermination().invoke(
+                        new Functions.TriConsumer<>() {
+                            @Override
+                            public void accept(Object o, Throwable throwable, Boolean cancelled) {
+                                recordCompletionResult(counted, commonTags, throwable);
+                            }
+                        });
             } catch (Throwable e) {
                 record(counted, commonTags, e);
             }

--- a/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/MicrometerTimedInterceptor.java
+++ b/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/MicrometerTimedInterceptor.java
@@ -18,6 +18,8 @@ import io.micrometer.core.instrument.MeterRegistry;
 import io.micrometer.core.instrument.Tags;
 import io.micrometer.core.instrument.Timer;
 import io.quarkus.arc.ArcInvocationContext;
+import io.smallrye.mutiny.Uni;
+import io.smallrye.mutiny.tuples.Functions;
 
 /**
  * Quarkus defined interceptor for types or methods annotated with {@link Timed @Timed}.
@@ -37,8 +39,8 @@ public class MicrometerTimedInterceptor {
     }
 
     @AroundInvoke
+    @SuppressWarnings("unchecked")
     Object timedMethod(ArcInvocationContext context) throws Exception {
-        final boolean stopWhenCompleted = CompletionStage.class.isAssignableFrom(context.getMethod().getReturnType());
         final List<Sample> samples = getSamples(context);
 
         if (samples.isEmpty()) {
@@ -46,17 +48,27 @@ public class MicrometerTimedInterceptor {
             return context.proceed();
         }
 
-        if (stopWhenCompleted) {
+        Class<?> returnType = context.getMethod().getReturnType();
+        if (TypesUtil.isCompletionStage(returnType)) {
             try {
                 return ((CompletionStage<?>) context.proceed()).whenComplete((result, throwable) -> {
-                    for (Sample sample : samples) {
-                        sample.stop(MicrometerRecorder.getExceptionTag(throwable));
-                    }
+                    stop(samples, MicrometerRecorder.getExceptionTag(throwable));
                 });
             } catch (Exception ex) {
-                for (Sample sample : samples) {
-                    sample.stop(MicrometerRecorder.getExceptionTag(ex));
-                }
+                stop(samples, MicrometerRecorder.getExceptionTag(ex));
+                throw ex;
+            }
+        } else if (TypesUtil.isUni(returnType)) {
+            try {
+                return ((Uni<Object>) context.proceed()).onTermination().invoke(
+                        new Functions.TriConsumer<>() {
+                            @Override
+                            public void accept(Object o, Throwable throwable, Boolean cancelled) {
+                                stop(samples, MicrometerRecorder.getExceptionTag(throwable));
+                            }
+                        });
+            } catch (Exception ex) {
+                stop(samples, MicrometerRecorder.getExceptionTag(ex));
                 throw ex;
             }
         }
@@ -68,9 +80,7 @@ public class MicrometerTimedInterceptor {
             exceptionClass = MicrometerRecorder.getExceptionTag(ex);
             throw ex;
         } finally {
-            for (Sample sample : samples) {
-                sample.stop(exceptionClass);
-            }
+            stop(samples, exceptionClass);
         }
     }
 
@@ -90,6 +100,12 @@ public class MicrometerTimedInterceptor {
             }
         }
         return samples;
+    }
+
+    private void stop(List<Sample> samples, String throwableClassName) {
+        for (int i = 0; i < samples.size(); i++) {
+            samples.get(i).stop(throwableClassName);
+        }
     }
 
     private void record(Timed timed, Timer.Sample sample, String exceptionClass, Tags timerTags) {
@@ -141,7 +157,7 @@ public class MicrometerTimedInterceptor {
         return Tags.of("class", className, "method", methodName);
     }
 
-    abstract class Sample {
+    abstract static class Sample {
 
         protected final Timed timed;
         protected final Tags commonTags;

--- a/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/TypesUtil.java
+++ b/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/TypesUtil.java
@@ -1,0 +1,19 @@
+package io.quarkus.micrometer.runtime;
+
+import java.util.concurrent.CompletionStage;
+
+import io.smallrye.mutiny.Uni;
+
+final class TypesUtil {
+
+    private TypesUtil() {
+    }
+
+    static boolean isUni(Class<?> clazz) {
+        return Uni.class.isAssignableFrom(clazz);
+    }
+
+    static boolean isCompletionStage(Class<?> clazz) {
+        return CompletionStage.class.isAssignableFrom(clazz);
+    }
+}

